### ISSUE TITLE
Topology should validate failureDomain before creating MachineDeployments

### ISSUE THREAD

**Issue by fabriziopandini** [kubernetes-sigs/cluster-api#5859]
**User Story**

As a user, I would like to have validation on `failureDomain` value selected for my MachineDeployments

**Detailed Description**

This is a follow up of https://github.com/kubernetes-sigs/cluster-api/pull/5850#discussion_r767892624

https://github.com/kubernetes-sigs/cluster-api/pull/5850 introduced the possibility to spread MachineDeployments across failureDomains, however, on creation it is not possible to validate the selected `failureDomain` matches one of the available failureDomains, given that they are derived from the InfrastructureCluster after Cluster creation.

So this issue is about discussing options to ensure that a MachineDeployment gets created only if the corresponding `failureDomain` exists.

Option 1. We should wait to create a MachineDeployment until available failureDomains are reported on the `Cluster.Status`, providing evidence of this in the TopologyReconciled conditions. If the required `failureDomain` does not match any available failureDomains, this is a reconcile error.

Option 2. If the available failureDomains are reported on the `Cluster.Status`, we should validate `failureDomain` for new MachineDeployments/edited MachineDeployments upfront.

Option 1 and 2 are not exclusive

/kind feature
/area topology
/milestone v1.1

**Comment by sbueringer** [comment]
I think if the failureDomain is set in MachineDeploymentTopology we should only create a MD when we can validate if the failureDomain is valid (i.e. if it's one of the failureDomains in the Cluster). 

Otherwise the machines might get created first in the wrong failure domain and then re-created once the Cluster contains the list of failureDomains.

**Comment by vincepri** [comment]
> The failureDomain is and should be mutable, right?

Yes, users can decide to move all machines in a MD to a different AZ and trigger a rollout

> I think if the failureDomain is set in MachineDeploymentTopology we should only create a MD when we can validate if the failureDomain is valid (i.e. if it's one of the failureDomains in the Cluster).

+1

**Comment by fabriziopandini** [comment]
moving out of the milestone because not staffed yet, but nice to have the soon as possible
/help
